src/Compilers/Core/Portable/Symbols/Attributes/CommonAttributeData.cs

```diff
@@ -292,9 +293,35 @@ private ObsoleteAttributeData DecodeObsoleteAttribute()
                 }
             }
 
-            return new ObsoleteAttributeData(ObsoleteAttributeKind.Obsolete, message, isError);
+            Debug.Assert(AttributeClass is object);
+
+            string? diagnosticId = null;
+            string? urlFormat = null;
+            foreach (var (name, value) in this.CommonNamedArguments)
+            {
+                if (diagnosticId is null && name == "DiagnosticId" && IsStringProperty("DiagnosticId"))
+                {
+                    diagnosticId = value.ValueInternal as string;
+                }
+                else if (urlFormat is null && name == "UrlFormat" && IsStringProperty("UrlFormat"))
+                {
+                    urlFormat = value.ValueInternal as string;
+                }
+
+                if (diagnosticId is object && urlFormat is object)
+                {
+                    break;
+                }
+            }
+
+            return new ObsoleteAttributeData(ObsoleteAttributeKind.Obsolete, message, isError, diagnosticId, urlFormat);
         }
 
+        // Note: it is disallowed to declare a property and a field
+        // with the same name in C# or VB source, even if it is allowed in IL.
+        // We use a virtual method and override to prevent having to realize the public symbols just to decode obsolete attributes.
+        private protected virtual bool IsStringProperty(string memberName) => throw ExceptionUtilities.Unreachable;
```

Comment:
abstract?

Reply:
Adding a comment to clarify why this is 'virtual' and not 'abstract'.